devicetypes/smartthings/rgbw-light.src/rgbw-light.groovy

```diff
@@ -232,53 +266,105 @@ def setHue(value) {
 
 def setColor(value) {
 	log.debug "setColor($value)"
-	def result = []
+	def rgb
+	if (state.staged == null) {
+		state.staged = [:]
+	}
 	if (value.hex) {
-		def c = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) }
-		result << zwave.switchColorV3.switchColorSet(red:c[0], green:c[1], blue:c[2], warmWhite:0, coldWhite:0)
+		state.staged << [color: value.hex] // stage ST RGB color attribute
+		def hsv = colorUtil.hexToHsv(value.hex) // convert to HSV
+		state.staged << [hue: hsv[0], saturation: hsv[1]] // stage ST hue and saturation attributes
+		rgb = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) } // separate RGB elements for zwave setter
 	} else {
-		def rgb = huesatToRGB(value.hue, value.saturation)
-		result << zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2], warmWhite:0, coldWhite:0)
+		state.staged << value.subMap("hue", "saturation") // stage ST hue and saturation attributes
+		def hex = colorUtil.hsvToHex(Math.round(value.hue) as int, Math.round(value.saturation) as int) // convert to hex
+		state.staged << [color: hex] // stage ST RGB color attribute
+		rgb = colorUtil.hexToRgb(hex) // separate RGB elements for zwave setter
 	}
-	if (device.currentValue("switch") != "on") {
-		result << zwave.basicV1.basicSet(value: 0xFF)
-		result << zwave.switchMultilevelV3.switchMultilevelGet()
+	commands([zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2], warmWhite: 0, coldWhite: 0),
+	          zwave.switchColorV3.switchColorGet(colorComponent: RGB_NAMES[0]), // event-publish callback is on any of the RGB responses, so only need to GET one of these
+	], 3500)
+}
+
+private emitTemperatureSet(temp, cmds) {
+	// Restrict temp to legal bounds.
+	temp = BOUND(temp, COLOR_TEMP_MIN, COLOR_TEMP_MAX)
+	// Add our staging dictionary to state.
+	if (state.staged == null) {
+		state.staged = [:]
 	}
-	result += queryAllColors()
+	// Stage ST colorTemperature attribute.
+	state.staged << [colorTemperature: temp]
+	// Emit our command.  Follow up with callback-trigger getter.  Our
+	// event-publish callback is on any of the while-level responses,
+	// so we only need to GET one of these these.  Make sure that we
+	// only have delay immediately before the callback trigger getter.
+	def prologue = cmds.init() // grab all but last
+	def epilogue = [] << cmds.last() // grab last
+	epilogue << zwave.switchColorV3.switchColorGet(colorComponent: WHITE_NAMES[0]) // append callback get
+	def rv = prologue.size() > 0 ? commands(prologue, 0) : [] // collect formatted prologue; check for empty; empty is OK, but can't be passed to commands()
+	rv << commands(epilogue, 3500) // collect formatted epilogue; only delay immediately before callback getter
+	return rv // return formatted command array to execute
```

Comment:
Just so that I understand this function, the goal is to take all of the temperature set commands, add a delay of 3500, and then add the get for the white color component? If so, `delayBetween` basically just puts `delay $delayValue` in between each item in the array. You can do this manually if you wanted. An example is here: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/1a19a83ce401baf2e6ed74975bbe3ed3d257a3bf/devicetypes/smartthings/zwave-motion-temp-light-sensor.src/zwave-motion-temp-light-sensor.groovy#L174

So you probably could have:
```
def result = [];

result += commands(cmds)
result += ["delay 3500", command(zwave.switchColorV3.switchColorGet(colorComponent: WHITE_NAMES[0]))]

return result
```

That said... It definitely feels like a bad idea to have that hardcoded around (which I am guilty of doing). Too bad we didn't just create a `delay(delayValue) { "delay $delayValue" }` API method for device types to use.

I'll leave it to your discretion if that is the case and if you want to change it. I should have paid a little more attention to it earlier when I left the last comments. My apologies.

Reply:
This is fair.  Your solution is simpler.  As you allude, only real advantage to the approach in the PR is that the delay command string is abstracted / obfuscated through the delayBetween API call.

Had I known before that this simple command string was available, I probably would have used it.  For now, I am going to leave this since it does work and I'd like to close out this work.

Thank you for the careful review.